src/ODEmodel.jl

```diff
@@ -44,80 +44,72 @@ end
 
 
 """ Predicts the model given a set of parametrs. """
-function predict(p::Vector{T}, g_0, t; g1data = nothing, g2data = nothing) where {T}
-    @assert length(p) == 9
+function predict(p, g_0, t)
     # Convert parameters to phase numbers
     nG1 = Int(floor(p[6]))
     nG2 = Int(floor(p[7]))
     nD1 = Int(floor(p[8]))
     nD2 = Int(floor(p[9]))
+    if nD1 == 0
+        D1 = Float64[]
+    else
+        D1 = zeros(nD1)
+    end
+    if nD2 == 0
+        D2 = Float64[]
+    else
+        D2 = zeros(nD2)
+    end
 
     if g_0 isa Real
-        if nD1 == 0
-            D1 = T[]
-        else
-            D1 = zeros(nD1)
-        end
-
-        if nD2 == 0
-            D2 = T[]
-        else
-            D2 = zeros(nD2)
-        end
-
-        g_0 = [ones(nG1) * p[5] * g_0 / nG1; ones(nG2) * (1.0 - p[5]) * g_0 / nG2; D1; D2]
+        v = [ones(nG1) * p[5] * g_0 / nG1; ones(nG2) * (1.0 - p[5]) * g_0 / nG2; D1; D2]
+    else
+        v = g_0
     end
 
     A = ODEjac(p, nG1, nG2, nD1, nD2)
 
-    prob = ODEProblem((du, u, p, t) -> mul!(du, A, u), g_0, maximum(t))
-    integrator = init(prob, VCABM(); save_on = false)
+    if t isa Real
+        v = expmv(t, A, v)
 
-    if g1data === nothing
-        G1 = Vector{T}(undef, length(t))
-        G2 = Vector{T}(undef, length(t))
+        G1 = sum(v[1:nG1]) + sum(v[(nG1 + nG2 + 1):(nG1 + nG2 + nD1)])
+        G2 = sum(v[(nG1 + 1):(nG1 + nG2)]) + sum(v[(nG1 + nG2 + nD1 + 1):(nG1 + nG2 + nD1 + nD2)])
     else
-        cost = 0.0
-    end
+        # Some assumptions
+        @assert t[1] == 0.0
+        rmul!(A, t[2])
+        A = SparseMatrixExp(A)
 
```

Comment:
Could you please let me know if these two functions for the sparse matrix are correctly here or not? It might be that these two are very slow or just loop forever.

Reply:
I'm not sure what you're trying to do? The exponential of a sparse matrix is typically not sparse, so I'm not sure how this function would exist. You could use Krylov solving with a sparse matrix, though. In that case you'd be calling something like expmv at each time point.